voting-contract/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::{env, near_bindgen, AccountId, Balance, BlockHeight, EpochHeight};
+use serde::Serialize;
+use std::collections::{BTreeMap, HashMap};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+const MIN_HEIGHT_HORIZON: u64 = 432000;
+const MAX_HEIGHT_HORIZON: u64 = MIN_HEIGHT_HORIZON * 100;
+
+type ProposalId = u64;
+
+#[derive(BorshDeserialize, BorshSerialize, Eq, PartialEq, Debug, Serialize)]
+pub struct Proposal {
+    /// Proposed height to reset the network.
+    proposed_height: BlockHeight,
+    /// Human-readable description of this proposal.
+    description: String,
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct AccountVotes {
+    votes: HashMap<ProposalId, Balance>,
+    total_voted_balance: Balance,
+    account_stake: Balance,
+}
+
+impl AccountVotes {
+    fn add_vote(&mut self, proposal_id: ProposalId, stake: Balance) {
+        self.total_voted_balance =
+            self.total_voted_balance + stake - *self.votes.get(&proposal_id).unwrap_or(&0);
+        self.votes.insert(proposal_id, stake);
+    }
+
+    fn withdraw_vote(&mut self, proposal_id: ProposalId) {
+        let stake = *self.votes.get(&proposal_id).unwrap_or(&0);
+        if self.total_voted_balance < stake {
+            env::panic(
+                format!(
+                    "trying to withdraw {} but only has {}",
+                    self.total_voted_balance, stake
+                )
+                .as_bytes(),
+            );
+        }
+        self.total_voted_balance -= stake;
+    }
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct Votes {
+    proposals: Map<ProposalId, Proposal>,
+    next_proposal_index: ProposalId,
+    proposed_heights: BTreeMap<BlockHeight, ProposalId>,
```

Comment:
I don't think you need BTreeMap for proposals. It all can be done offline and later the contract can be notified.

Reply:
Not sure what you mean. This is done so that it is easy to remove expired proposals.

Reply:
I suggest simplify this contract and don't keep things sorted internally. We can also don't need to automatically resolve proposals when the new vote arrives. Instead we can add a method `resolve` that takes a proposal ID. And it tries to resolve a given proposal. This way the contract internally doesn't have to garbage collect expired proposals, but instead rely on someone calling `resolve` on a given proposal.

Reply:
Okay sounds good.